File: src/main/java/org/irislang/jiris/core/IrisContextEnvironment.java
package org.irislang.jiris.core;

import java.util.HashMap;

public class IrisContextEnvironment {
	
	public enum RunTimeType {
		ClassDefineTime,
		ModuleDefineTime,
		InterfaceDefineTime,
		RunTime,
	}
	
	private IrisRunningObject m_runningType = null;
	private RunTimeType m_runTimeType = RunTimeType.RunTime; 
	
	private HashMap<String, IrisValue> m_localVariableMap = new HashMap<String, IrisValue>();
	private IrisContextEnvironment m_upperContext = null;
	
	private IrisMethod m_currentMethod = null;

	private IrisObject m_closureBlockObj = null;

    public IrisObject getClosureBlockObj() {
        return m_closureBlockObj;
    }
    
    public void setClosureBlockObj(IrisObject closureBlockObj) {
        this.m_closureBlockObj = closureBlockObj;
    }

    public IrisMethod getCurrentMethod() {
		return m_currentMethod;
	}
	
	public void setCurrentMethod(IrisMethod currentMethod) {
		m_currentMethod = currentMethod;
	}
	
	public IrisValue GetLocalVariableWithinChain(String localName) {
		IrisContextEnvironment tmp = this;
		IrisValue value = null;
		while(tmp != null) {
			value = tmp.GetLocalVariable(localName);
			if(value != null) {
				break;
			}
			tmp = tmp.m_upperContext;
		}
		return value;
	}
		
	public IrisValue GetLocalVariable(String localName) {
		return m_localVariableMap.get(localName);
	}
	
	public void AddLocalVariable(String localName, IrisValue value) {
		m_localVariableMap.put(localName, value);
	}
	
	public IrisRunningObject getRunningType() {
		return m_runningType;
	}
	
	public void setRunningType(IrisRunningObject runningType) {
		m_runningType = runningType;
	}

	public RunTimeType getRunTimeType() {
		return m_runTimeType;
	}

	public void setRunTimeType(RunTimeType runTymeType) {
		m_runTimeType = runTymeType;
	}
	
	public void setUpperContext(IrisContextEnvironment upperContext) {
		m_upperContext = upperContext;
	}
	
	public IrisContextEnvironment getUpperContext() {
		return m_upperContext;
	}
}

File: src/main/java/org/irislang/jiris/core/IrisContextEnvironment.kt
package org.irislang.jiris.core

import java.util.HashMap

class IrisContextEnvironment {

    var runningType: IrisRunningObject? = null
    var runTimeType = RunTimeType.RunTime

    private val m_localVariableMap = HashMap<String, IrisValue>()
    var upperContext: IrisContextEnvironment? = null

    var currentMethod: IrisMethod? = null

    var closureBlockObj: IrisObject? = null

    enum class RunTimeType {
        ClassDefineTime,
        ModuleDefineTime,
        InterfaceDefineTime,
        RunTime
    }

    fun GetLocalVariableWithinChain(localName: String): IrisValue? {
        var tmp: IrisContextEnvironment? = this
        var value: IrisValue? = null
        while (tmp != null) {
            value = tmp.GetLocalVariable(localName)
            if (value != null) {
                break
            }
            tmp = tmp.upperContext
        }
        return value
    }

    fun GetLocalVariable(localName: String): IrisValue? {
        return m_localVariableMap[localName]
    }

    fun AddLocalVariable(localName: String, value: IrisValue) {
        m_localVariableMap[localName] = value
    }
}
